File: source-connector/src/main/java/sourceconnector/domain/connect/FileSourceTask.java
package sourceconnector.domain.connect;

import lombok.AccessLevel;
import lombok.Getter;
import offsetmanager.domain.file.FileKey;
import offsetmanager.domain.offset.DefaultOffsetRecord;
import offsetmanager.domain.offset.OffsetStatus;
import sourceconnector.domain.log.Log;
import sourceconnector.domain.log.LogMetadata;
import sourceconnector.domain.pipeline.Pipeline;
import sourceconnector.domain.pipeline.factory.PipelineSupplier;
import sourceconnector.service.batcher.LogBatcher;
import sourceconnector.service.producer.BatchProducer;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FileSourceTask implements Task<FileProcessingResult> {
  private final String id;
  @Getter
  private final int index;
  private final BatchProducer<String> producer;
  private final PipelineSupplier<Log> pipelineSupplier;

  // visible for test
  @Getter(AccessLevel.PACKAGE)
  private final Map<FileKey, Long> fileKeyOffsetMap = new HashMap<>();
  private final FileProcessingResult result = new FileProcessingResult();

  public FileSourceTask(
    int index,
    PipelineSupplier<Log> pipelineSupplier,
    BatchProducer<String> producer
  ) {
    this.id = String.format("Task-%d", index);
    this.index = index;
    this.pipelineSupplier = pipelineSupplier;
    this.producer = producer;
  }

  @Override
  public FileProcessingResult call() throws Exception {
    try {
      for (Map.Entry<FileKey, Long> entry: this.fileKeyOffsetMap.entrySet()) {

        long offset = entry.getValue();
        if (offset == OffsetStatus.COMPLETED.getValue()) {
          result.addSkippedCount();
          continue;
        }
        FileKey fileKey = entry.getKey();
        Pipeline<Log> pipeline = pipelineSupplier.get(fileKey);

        // Progress offset to the next position in the file
        pipeline.toPosition(offset);

        LogBatcher batcher = new LogBatcher(pipeline, 10_000);

        LogMetadata lastMessageMetadata;

        while (batcher.hasNextBatch()) {
          List<Log> messages = batcher.nextBatch().get();
          if (messages.isEmpty()) continue;
          lastMessageMetadata = messages.getLast().getMetadata();
          List<String> messageBatch = messages
            .stream()
            .map(Log::get)
            .toList();

          producer.sendBatch(
            new DefaultOffsetRecord(
              lastMessageMetadata.key(),
              lastMessageMetadata.offset()
            ),
            ()->messageBatch
          );
        }


        // Complete this file
        producer.sendBatch(
          new DefaultOffsetRecord(
            // This is for handling no Log after filtered
            fileKey,
            OffsetStatus.COMPLETED.getValue()
          ),
          Collections::emptyList
        );

        this.result.addSuccessCount();
      }

      return this.result;
    } finally {
      this.producer.close();
    }
  }

  @Override
  public void assign(Map<FileKey, Long> fileKeys) {
    this.fileKeyOffsetMap.putAll(fileKeys);
    this.result.setTotalCount(fileKeyOffsetMap.size());
  }
}

File: source-connector/src/main/kotlin/sourceconnector/domain/connect/FileSourceTask.kt
package sourceconnector.domain.connect

import offsetmanager.domain.file.FileKey
import offsetmanager.domain.offset.DefaultOffsetRecord
import offsetmanager.domain.offset.OffsetStatus
import org.slf4j.LoggerFactory
import sourceconnector.domain.log.Log
import sourceconnector.domain.pipeline.factory.PipelineSupplier
import sourceconnector.service.batcher.LogBatcher
import sourceconnector.service.producer.BatchProducer
import sourceconnector.service.producer.ProduceResult

class FileSourceTask(
  override val index: Int,
  private val pipelineSupplier: PipelineSupplier<Log>,
  private val producer: BatchProducer<String>,
  private val id: String = String.format("Task-%d", index)
) : Task<FileProcessingResult> {

  private val log = LoggerFactory.getLogger(FileSourceTask::class.java)
  // visible for test
  internal val fileKeyOffsetMap: MutableMap<FileKey, Long> = mutableMapOf()
  private lateinit var result: FileProcessingResult

  @Throws(Exception::class)
  override fun call(): FileProcessingResult {
    try {
      for ((fileKey, offset) in this.fileKeyOffsetMap) {
        if (offset == OffsetStatus.COMPLETED.offset) {
          result.addSkippedCount()
          continue
        }

        if (processFile(fileKey, offset).isFailure) {
          log.error("Failed to produce file: {}", fileKey.get())
          this.result.addFailure(fileKey)
        } else {
          this.result.addSuccessCount()
        }
      }

      return this.result
    } finally {
      this.producer.close()
    }
  }

  override fun assign(offsetMap: Map<FileKey, Long>) {
    this.fileKeyOffsetMap.putAll(offsetMap)
    this.result = FileProcessingResult(offsetMap.size)
  }

  private fun processFile(fileKey: FileKey, offset: Long): ProduceResult {
    pipelineSupplier.get(fileKey).use { pipeline ->
      pipeline.toPosition(offset)
      val batcher = LogBatcher(pipeline, 10000)

      while (batcher.hasNextBatch()) {
        val messages = batcher.nextBatch().get()
        if (messages.isEmpty()) continue

        val lastMetadata = messages.last().metadata
        val batch = messages.map { it.get() }

        val result: ProduceResult = producer.sendBatch(
          DefaultOffsetRecord(lastMetadata.key, lastMetadata.offset)
        ) { batch }

        // Early return if fail before completing process
        if (result.isFailure) return result
      }

      return producer.sendBatch(
        DefaultOffsetRecord(fileKey, OffsetStatus.COMPLETED.offset)
      ) { emptyList() }
    }
  }
}
